File: SearchinRotatedSortedArrayII.py
class Solution:
    def search(self, nums, target):
        """
        :type nums: List[int]
        :type target: int
        :rtype: bool
        """
        l = 0
        r = len(nums)-1
        while l<=r:
            mid = (l+r)//2
            if nums[mid]==target:
                return True
            elif nums[mid]==nums[l] and nums[mid]==nums[r]:
                l+=1
                r-=1
            elif nums[mid]>=nums[l]:
                if nums[mid]>target and nums[l]<=target:
                    r = mid-1
                else:
                    l = mid+1
            else:
                if nums[mid]<target and nums[r]>=target:
                    l = mid+1
                else:
                    r = mid-1
        return False
        
if __name__=="__main__":
    nums=list(input())
    target=input()
    solution=Solution()
    print(solution.search(nums, target))

File: SearchinRotatedSortedArrayII.cpp
#include<iostream>
#include<vector>

using namespace std;

class Solution {
public:
    bool search(vector<int>& nums, int target) {
        int l=0, r=nums.size()-1;
        
        while(l<=r){
            int mid=(l+r)/2;
            if(nums[mid]==target) return true;
            
            else if(nums[mid]>nums[r]){
                if(nums[mid]>target&&nums[l]<=target){
                    r=mid;
                }else{
                    l=mid+1;
                }
            }else if(nums[mid]<nums[r]){
                if(nums[mid]<=target&&nums[r]>=target){
                    l=mid+1;
                }else{
                    r=mid;
                }
            }else{
                r--;
            }
            
            cout<<"l: "<<l<<endl;
            cout<<"r: "<<r<<endl;
        }
        
        return false;
    }
};

int main(){
    int n;
    cin>>n;
    
    vector<int> nums;
    for(int i=0;i<n;i++){
        int num;
        cin>>num;
        nums.push_back(num);
    }
    
    int target;
    cin>>target;
    
    Solution *solution=new Solution();
    solution->search(nums, target) ? cout<<"True" : cout<<"False";
    
    return 0;
}
